pkg/backends/calico/routes.go

```diff
@@ -190,29 +190,156 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 
 	// see if any endpoints are on this node and advertise if so
 	// else remove the route if it also already exists
 	rg.Lock()
-	if rg.advertiseThisService(svc, ep) {
-		route := svc.Spec.ClusterIP + "/32"
-		if cur, exists := rg.svcRouteMap[key]; !exists {
-			// This is a new route - send it through.
-			rg.advertiseRoute(key, route)
-		} else if route != cur {
-			// The route has changed somehow. Send a delete for the old one
-			// and add the new one. We don't expect to hit this, since cluster IPs
-			// are immutable. But, handle it for good measure.
-			logc.Warn("ClusterIP for service changed, adjusting")
-			rg.withdrawRoute(key, cur)
+	defer rg.Unlock()
+
+	advertise := rg.advertiseThisService(svc, ep)
+	if advertise {
+		routes := rg.getAllRoutesForService(svc)
+		rg.advertiseRoutes(key, routes)
+	} else {
+		routes := rg.getAdvertisedRoutes(key)
+		rg.withdrawRoutes(key, routes)
+	}
+
+}
+
+// onBGPConfigurationUpdate is called from the calico client, whenever there
+// is a change to the svc_external_ips, and will update all of the routes
+// advertised for each service.
+func (rg *routeGenerator) onBGPConfigurationUpdate() {
+
+	// Get all the services that we know about
+	svcs := make([]*v1.Service, 0)
+	svcIfaces := rg.svcIndexer.List()
+	for _, svcIface := range svcIfaces {
+		svc, ok := svcIface.(*v1.Service)
+		if !ok {
+			log.Error("Type assertion failed for rg.svcIndexer result member. Will not process updates to routes advertised for service.")
+			continue
+		}
+
+		svcs = append(svcs, svc)
+	}
+
+	// Update the routes advertised for each service
+	for _, svc := range svcs {
+		rg.setRouteForSvc(svc, nil)
+	}
+
+}
+
+// getAllRoutesForService returns all the routes that should be advertised
+// for the given service.
+func (rg *routeGenerator) getAllRoutesForService(svc *v1.Service) []string {
+
+	routes := make([]string, 0)
+	routes = append(routes, svc.Spec.ClusterIP)
+
+	if svc.Spec.ExternalIPs != nil {
+		for _, externalIP := range svc.Spec.ExternalIPs {
+
+			// Only advertise whitelisted external IP's
+			if !rg.isAllowedExternalIP(externalIP) {
+				log.Infof("External IP not advertised as not whitelisted: %s", externalIP)
+				continue
+			}
+
+			routes = append(routes, externalIP)
+		}
+	}
+
+	return addSuffix(routes, "/32")
+
+}
+
+// getAdvertisedRoutes returns the routes that are currently advertised and
+// associated with the given key.
+func (rg *routeGenerator) getAdvertisedRoutes(key string) []string {
+
+	routes := make([]string, 0)
+
+	if rg.svcRouteMap[key] != nil {
+		for route := range rg.svcRouteMap[key] {
+			routes = append(routes, route)
+		}
+	}
+
+	return routes
+
+}
+
+// advertiseRoutes associates only the given routes with the given key,
+// and advertises the given routes. It also withdraws any routes that are no
+// longer associated with the given key.
+func (rg *routeGenerator) advertiseRoutes(key string, routes []string) {
```

Comment:
minor point: I'd recommend calling this `setRoutes` to distinguish it from `advertiseRoute` - since this is a bit more sophisticated than simply advertising the given routes (it also withdraws any that should no longer be there)